mvebu: add support for Western Digital My Cloud EX2 Ultra NAS

evgenyz opened the pull request:
Device:

- Marvell Armada 385 (2x1.33GHz) / Marvell 88F6820
- 1x 10/100/1000 Mbps Ethernet (with WOL support)
- 2x USB3 ports
- 2x SATA ports (2x 3.5" HDD slots enclosure)
- 256 MB of Flash
- 1 GB of RAM
- UART
- 1x Button (reset)

Installation (initial):

- Using original firmware's network settings obtain SSH access to the device.
- Put *-image.cfs-factory.img and *-uImage-factory.img images into device's /tmp directory.
- Write kernel (uImage) image 'flash_eraseall /dev/mtd1 && nandwrite --markbad -p /dev/mtd1 /tmp/*-uImage-factory.img'.
- Write rootfs (image.cfs) image 'ubiformat /dev/mtd3 -f /tmp/*-image.cfs-factory.img -y'.
- Reboot the device.

Installation (upgrade):

Use *-sysupgrade.bin in a usual way.

Restoring original firmware:

- Erase kernel partition 'mtd erase mtd1'.
- Reboot the device (it will fall into the recovery mode).
- Upload original firmware image using recovery's web interface.

Features:

Indicators, network, storage, termal control, WOL, reset button. Pretty much all
of the basic hardware components are stable and functional.

Known bugs:

- mdadm tends to crash ubifs driver in default configuration.

[boot-first.log](https://github.com/openwrt/openwrt/files/3154890/boot-first.log)
[boot-sysupgrade.log](https://github.com/openwrt/openwrt/files/3154891/boot-sysupgrade.log)
[boot-wd-original.log](https://github.com/openwrt/openwrt/files/3154892/boot-wd-original.log)

evgenyz committed:
mvebu: add support for Western Digital My Cloud EX2 Ultra NAS

Device:

- Marvell Armada 385 (2x1.33GHz) / Marvell 88F6820
- 1x 10/100/1000 Mbps Ethernet (with WOL support)
- 2x USB3 ports
- 2x SATA ports (2x 3.5" HDD slots enclosure)
- 256 MB of Flash
- 1 GB of RAM
- UART
- 1x Button (reset)

Installation (initial):

- Using original firmware's network settings obtain SSH access to the device.
- Put *-image-cfs-factory.bin and *-uImage-factory.bin images into device's /tmp directory.
- Write kernel (uImage) image 'flash_eraseall /dev/mtd1 && nandwrite --markbad -p /dev/mtd1 /tmp/*-uImage-factory.bin'.
- Write rootfs (image-cfs) image 'ubiformat /dev/mtd3 -f /tmp/*-image.cfs-factory.bin -y'.
- Reboot the device.

Installation (upgrade):

Use *-sysupgrade.bin in a usual way.

Restoring original firmware:

- Erase kernel partition 'mtd erase mtd1'.
- Reboot the device (it will fall into the recovery mode).
- Upload original firmware image using recovery's web interface.

Features:

Indicators, network, storage, termal control, WOL, reset button.
Pretty much all of the basic hardware components are stable
and functional.

Known bugs:

- mdadm tends to crash ubifs driver in default configuration.

Signed-off-by: Evgeny Kolesnikov <evgenyz@gmail.com>

evgenyz commented:
New package wtmcu: openwrt/packages#8929

diizzyy commented:
Why is a new subtarget needed?

Are AHCI packages needed? https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/config-4.14#L1

Are USB-related packages needed? https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/config-4.14#L474

I think in general that you should refrain from including possibly unneeded packages/lkmods in general, see: https://github.com/openwrt/openwrt/blob/master/target/linux/apm821xx/sata/target.mk for example.

The wtmpcu package needs to go into base instead of packages

diizzyy commented on target/linux/mvebu/base-files/etc/board.d/02_network:
```diff
@@ -28,6 +28,9 @@ armada-385-turris-omnia)
 	ucidef_set_interface_wan "eth2"
 	;;
+armada-385-wd-mcex2u)
+	ucidef_set_interface_lan "eth0" "dhcp"
```
I'm in favour of this as it "makes sense" but looking https://github.com/openwrt/openwrt/pull/1062 it's still a mess....

evgenyz replied:
Well, I checked the thread. While, in general, I tend to agree with your approach to make one-port devices DHCP-driven by default, I could see at least one somewhat valid point against: OpenWrt does require configuring after installation and looking for the device could be problematic. But on the other hand I could see static approach to be even more problematic in case there is already an OpenWrt router serving such one-port device like NAS (easily imaginable situation, actually). 

Still, I think that, regardless of overall stance on "dhcp by default for one-ports" is being undefined, using it on a particular device, with a known upgrade workflow, should be acceptable.

Like it is in this case — upgrade procedure could be done via SSH on the original firmware, user is aware of the IP, MAC address is properly inherited by the OpenWrt firmware after upgrade. So, considering the principle of least surprise, I'd prefer to leave as it is.

evgenyz commented:
I'm going to join wtmcu package here and address your proposals in the nearest future

ynezz commented on target/linux/mvebu/Makefile:
```diff
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 BOARDNAME:=Marvell EBU Armada
 FEATURES:=fpu usb pci pcie gpio nand squashfs ramdisk boot-part rootfs-part
-SUBTARGETS:=cortexa9 cortexa53 cortexa72
+SUBTARGETS:=cortexa9 cortexa53 cortexa72 wdmc
```
Why do you introduce this new subtarget?

evgenyz replied:
My reasoning was pretty much the same as in https://github.com/openwrt/openwrt/tree/master/target/linux/apm821xx with nand and sata subtargets: to isolate WD NAS-related stuff and Weltrend MCU related quirks. While there are at least 3 more devices like My Cloud Ultra I'd like work on and add under this subtarget I don't know if it's a good idea to group them into something generic like "sata" or "nas".

ynezz replied:
I'm probably confused, because you're adding `wdmc` subtarget, but then you're still adding `Device/armada-385-wd-mcex2u` into `cortexa9` subtarget and into `wdmc` subtarget as well.

> My reasoning was pretty much the same as in https://github.com/openwrt/openwrt/tree/master/target/linux/apm821xx with nand and sata subtargets: 

You still didn't convinced me, that support for this device shouldn't go into `cortexa9` subtarget.

> to isolate WD NAS-related stuff and Weltrend MCU related quirks

I don't see any of these in this PR, can you be more specific?

evgenyz replied:
> I'm probably confused, because you're adding `wdmc` subtarget, but then you're still adding `Device/armada-385-wd-mcex2u` into `cortexa9` subtarget and into `wdmc` subtarget as well.

I'm sorry, I just missed it while I was working on moving it into subtarget.

> 
> > My reasoning was pretty much the same as in https://github.com/openwrt/openwrt/tree/master/target/linux/apm821xx with nand and sata subtargets:
> 
> You still didn't convinced me, that support for this device shouldn't go into `cortexa9` subtarget.
> 
> > to isolate WD NAS-related stuff and Weltrend MCU related quirks
> 
> I don't see any of these in this PR, can you be more specific?

Things like wtmcu package itself, wdmc/base-files/etc/diag.sh. And it might be more hairy in the future as other devices uses different MCU versions and indication logic.

Also I've used ARCH_WANT_LIBATA_LEDS, which does not do any good for devices without dedicated indicator for each drive. (It's still in mvebu though, in form of a patch as I don't understand how to isolate it into a subtarget properly. Need some advice here).

ynezz replied:
> Things like wtmcu package itself

Isn't `DEVICE_PACKAGES += wtmcu` enough?

> wdmc/base-files/etc/diag.sh

I don't see a point of having separate subtarget just to use different diag.sh, you should probably extend current diag.sh in order to support your use case

> also I've used ARCH_WANT_LIBATA_LEDS, which does not do any good for devices without dedicated indicator for each drive.

Ideally this patch should be replaced with some upstreamable solution, but until then, I see no problem in adding this config option to `cortexa9` subtarget as I think, that kernel size and runtime overhead is negligible on this platform.

evgenyz replied:
Merged device into `cortexa9` subtarget.

ynezz commented on target/linux/mvebu/base-files/etc/board.d/01_leds:
```diff
@@ -41,6 +41,12 @@ armada-385-linksys-venom)
 	ucidef_set_led_usbport "usb2_ss" "USB 2 SS" "pca963x:venom:blue:usb3_2" "usb3-port1"
 	;;
+armada-385-wd-mcex2u)
+	# ucidef_set_led_default "hdd1" "HDD1 Activity" "wdmcex2u:blue:hdd1" "1"
```
If you don't need those, then remove it, but don't keep this commented out code.

ynezz commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+/*
```
This DTS should be submitted to the upstream kernel first, then backported to OpenWrt tree.

ynezz commented on target/linux/mvebu/image/cortex-a9.mk:
```diff
@@ -90,6 +90,23 @@ define Device/armada-385-db-ap
 TARGET_DEVICES += armada-385-db-ap
 
+define Device/armada-385-wd-mcex2u
+#  $(Device/NAND-128K)
```
Please take a rid of this commented out code.

ynezz commented on target/linux/mvebu/patches-4.14/530-mvebu-wdmc-poweroff-restart.diff:
```diff
@@ -0,0 +1,220 @@
+ obj-$(CONFIG_POWER_RESET_VERSATILE) += arm-versatile-reboot.o
+diff -u -r -N linux-4.14.18.orig/drivers/power/reset/wt-poweroff.c linux-4.14.18/drivers/power/reset/wt-poweroff.c
+--- linux-4.14.18.orig/drivers/power/reset/wt-poweroff.c	1970-01-01 01:00:00.000000000 +0100
++++ linux-4.14.18/drivers/power/reset/wt-poweroff.c	2018-02-11 21:17:35.129989764 +0100
```
This driver should go through upstream first, then eventually backported to the OpenWrt in target/linux/generic/backport-y.xx directory.

ynezz commented on target/linux/mvebu/wdmc/base-files/lib/preinit/05_wdmc_config:
```diff
@@ -0,0 +1,21 @@
+		damaged=0
+		syscfg_part=$(grep config /proc/mtd |cut -c4)
+		ubiattach -m $syscfg_part -d 1 2>/dev/null || damaged=1
+		mkdir /tmp/wd-config
```
you're reinventing the wheel here, just use (or extend) target/linux/mvebu/base-files/lib/preinit/81_linksys_syscfg

evgenyz replied:
It's almost just a copy of 81_linksys_syscfg, I did it only because this partition is needed by 06_set_iface_mac, 81 is way too late.

ynezz replied:
then perhaps your `06_set_iface_mac` is wrong and you should do there something like `mac=$(mtd_get_mac_ascii config mac_addr)` instead of this weird `/tmp/wd-config` fiddling?

evgenyz replied:
The closest thing is mtd_get_mac_binary_ubi. Which is not really suitable, because there are evidences of the whole partition content being rewritten under some conditions in the original firmware (not all parts are open), and even in uboot they mount the partition to get the MAC. Reading it from a constant offset seems to be unreliable.

evgenyz replied:
Dead end (see below)

ynezz commented on target/linux/mvebu/base-files/lib/mvebu.sh:
```diff
@@ -17,6 +17,9 @@ mvebu_board_detect() {
 	*"Marvell Armada 370 Evaluation Board")
 		name="armada-370-db"
 		;;
+	*"WD My Cloud EX2 Ultra")
```
This reminds me of this [patch series](https://patchwork.ozlabs.org/project/openwrt/list/?series=105587) from @tmn505 where with this [patch](https://patchwork.ozlabs.org/patch/1093770/) we could start using DT based board detection so this could be dropped.

evgenyz replied:
Patch is in master and PR is adopted to it

ynezz commented on target/linux/mvebu/wdmc/base-files/lib/preinit/05_wdmc_leds:
```diff
@@ -0,0 +1,20 @@
+	armada-385-wd-mcex2u)
+		[ -e /dev/sda ] && {
+			echo "HDD1 (sda) is online, LED trigger ata1" > /dev/kmsg
+			led_set_attr "wdmcex2u:blue:hdd1" "trigger" "ata1"
```
This LED triggers should be set in the DTS via `linux,default-trigger = "ata1";` property. Well, ideally with upstream supported trigger `linux,default-trigger = "disk-activity";` but AFAIK this trigger currently doesn't support different/selectable ports.

evgenyz replied:
It's a bit tricky here. These blue LEDs are used as an activity as well as disk presence/healthy state indicators. They are inverted. So just defining ataX in DTS is not enough, LEDs will always be ON, even for absent drives. Also, as they share physical space with the red ones (not inverted), they should be turned off when we want indicate an error (like SMART failure, feature that I'm still working on).

Yes, disk-activity is only suitable for a single HDD LED at this moment, hence is all that ARCH_WANT_LIBATA_LEDS jumping around.

ynezz commented on target/linux/mvebu/image/wdmc.mk:
```diff
@@ -0,0 +1,21 @@
+
+  KERNEL_SIZE := 5120k
+  DEVICE_TITLE := WD My Cloud EX2 Ultra (DB-88F6820)
+  DEVICE_PACKAGES += wtmcu -dnsmasq -ppp
```
BTW where is this wtmcu package? I can't find it in this PR or in the OpenWrt main/packages trees.

evgenyz replied:
Here: openwrt/packages#8929. I'm moving it into this PR, as was suggested.

ynezz replied:
Well I didn't suggested this, I was just wondering where it is as I've never heard about it before and couldn't find it either. Anyway, as you can't depend in the image on the package which is not part of OpenWrt main tree, it should be included as part of this PR, so whoever suggested it, was mostly correct.

By mostly above I mean, that I've just quickly looked at source code of that wtmcu package and it seems to me, that most of this functionality belongs to the kernel instead to the userspace application. It should be probably implemented in the same spirit as your current `wdmcex2u,power-off` so you would need implement driver for `wdmcex2u,leds`, `wdmcex2u,fan` and `wdmcex2u,hwmon` (just my very rough estimation and upstream might see it differently).

You can ask upstream developers for suggestions once you start upstreaming that `wdmcex2u,power-off` driver and/or the DTS for this device.

evgenyz replied:
I've considered this approach. And while fan and hwmon related functions are feasible as a driver, things like WOL (is not enough to just prime phy device, MCU has another switch), wake-up alarm and LEDs (fixed modes and patterns, no way to control brightness, tty exchange speed won't allow them to really 'blink' in a custom way) are not so fitted into driver models as far I can imagine. 

Also, I could not find any intention from WD or Weltrend to make a proper driver for this MCU or at least open it's specification. I'm afraid I won't have enough resources to implement it all by myself.

The wdmcex2u,power-off is really a simple hack and pretty much essential to be a driver for the board to be adequately functional (even then WD uses userspace tool for that, lot's of WOL bugs related to improper shutdown, their firmware is an abomination). I'm not sure if upstream will accept a bunch hacks like that applied to variety of drivers like net/phy/marvell as they are very board-specific.

evgenyz commented:
@ynezz 
We are having pretty intense exchange about a lot of things, some are easily fixable/changeable and some are not so much. I would like to clarify for myself and summarize overall state of this PR. 

As far as I understand this device could not be added to the OpenWrt until I send (and kernel upstream accept) patches with the DTS of the device and (at least) wt-poweroff-reset driver.

And it is not acceptable to have MCU control in a form of a userspace daemon and all its functions should be implemented as drivers (or dropped if it is not feasible) and integrated in kernel as well, right?

And if kernel won't accept it for some reason — there is no way for the device to get into the OpenWrt distribution, right?

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
```
Instead of this text, add SPDX license identifier at the beginning of the file.

evgenyz replied:
Loading file '/mac_addr' to addr 0x02000000 with size 18 (0x00000012)...
They load it into memory and then clearly use it for initializing eth because 'dhcp' in u-boot sends correct MAC. But I can't see it anywhere in 'env print -a'. And then they use mounted config partition later in the boot process for getting the address.

tmn505 replied:
Can You check if adding
````
aliases {
        ethernet0 = &eth0;
};
````
node (below `chosen`) will work? We do this for ESPRESSObin boards and it uses same network driver, so theoretically should work.

evgenyz replied:
Nope, no cigar. Might it require some property override, like memory address?

evgenyz replied:
One thing has changed though:
[    0.953300] mvneta f1034000.ethernet eth0: Using random mac address 96:88:08:66:ae:5d
Became
[    1.007963] mvneta f1034000.ethernet eth0: Using hardware mac address 00:50:43:f6:1a:24
But the MAC is wrong

tmn505 replied:
Can You try changing `ethernet0` to `ethernet1` and after `ethernet2`?

evgenyz replied:
Nope, ethernet1 and ethernet2 changes nothing. Can you elaborate what are we looking here for? Where should U-Boot place the MAC? Maybe I could just write a bootscript for this.

For some reason I can't attach environment log here. I'll try to put it elsewhere.

tmn505 replied:
The `ethernet0 = &eth0` tells the U-Boot to replace MAC address of eth0 with data from ethernet0. In particular when ethernet0 is specified it takes MAC from ethaddr or eth0addr variable, when ethernet1 it should take address from eth1addr, when ethernet2 from eth2addr and so on. I thought that maybe U-Boot stored the proper MAC in ethXaddr value, so that's why i asked about next tests. You can try it out, set dts back to "ethernet0 = &eth0" and change the value of ethaddr to Your MAC, save Your environment and reboot. It might not work because this MAC:
`mvneta f1034000.ethernet eth0: Using hardware mac address 00:50:43:f6:1a:24`
isn't the same as value of ethaddr in environment You posted.
As I look more on the issue, it seems like Your approach with setting MAC is a proper one.

evgenyz replied:
So, saving real MAC to `eth2addr` in U-Boot does the trick. But this U-Boot version lacks `setexpr` and I have no other idea about how to write MAC, which is loaded into memory from UBI partition, into the environment variable. I guess bootscript it's really a dead end.

bobafetthotmail replied:
can this uboot save the environment vars with **saveenv**? 

I see the "       _U-Boot Environment:      0x00000000:0x00080000 (NAND)_" in the boot logs so I think there is a writable environment.

Also check if you have "*setenv*", on Kirkwood devices (Marvell hardware) the uboot has "*setenv*" command to set env variables.

Also if you can't write it from uboot you can write it from OpenWrt with the tool in the uboot-envtools package (if you can't automate it you can ask the user to do this as part of the installation procedure)

evgenyz replied:
There is no *saveenv*. While *setenv* is present, and does the trick when I manually save right MAC into the variable (I guess saveenv is just not needed). But we are looking for an automated (and more elegant than mounting config partition in preinit) way to do this.

tmn505 replied:
> So, saving real MAC to eth2addr in U-Boot does the trick.

That's great, with that we can do something.

> But this U-Boot version lacks setexpr and I have no other idea about how to write MAC, which is loaded into memory from UBI partition, into the environment variable.

We can use for that uboot-envtools, instead of fiddling with U-Boot command line and bootscripts. The idea would be, add alias node to dts; on first boot, in preinit script, read out MAC from config partition; set the eth2addr variable and additional one like for example mac_set=1.  The additional variable would be for testing if the U-Boot environment was reset, if yes script will redo whole procedure. That way we won't mount config partition on every boot, it should speed up the boot process a bit.

BTW, in `u-boot-2013.01-2014_T3.0p2/board/mv_ebu/a38x/mv_main_a38x.c` from WDMyCloud_Ex2Ultra_GPL_v2.31.163_20181219 tarball is where the "magic" happens regarding MAC address, that is also a reason why I asked for U-Boot environment to check if anything is writing to `0x02000000` and as it stands, yes it does, so we couldn't read it from there.

evgenyz replied:
OK. I surrender.

First of all, after a power cycle manually set `eth2addr` disappears, which is expectable as I did not save it with `saveenv` (because it is absent). Then I tried to locate where these 0x80000 of env are actually stored. The answer is: *nowhere*.

They define it's offset as `#define CONFIG_ENV_OFFSET nand_get_env_offs()`, which boils down to `offset = CONFIG_UBOOT_SIZE + CONFIG_SPARE_AREA` or somewhere further away if `CONFIG_SKIP_BAD_BLOCK` is defined.

Given `#define CONFIG_UBOOT_SIZE   0x100000` and `#define CONFIG_SPARE_AREA   0x400000`, we have offset no less than 0x500000. But the whole size of the U-Boot partition is 0x500000, next, without any gaps, goes kernel. Huh?

evgenyz replied:
Screw that. They actually define is as 0 somewhere, but it does not matter. They use environment as local variables, not storing them anywhere and not reading them from anywhere, as far as I can see from their U-Boot sources.

tmn505 replied:
> Screw that. They actually define it as 0 somewhere, but it does not matter. They use environment as local variables, not storing them anywhere and not reading them from anywhere, as far as I can see from their U-Boot sources.

True, that was my derp moment, sorry for pushing You down the rabbit hole. Indeed the U-Boot recreates the environment (with values from the file I mentioned) on each reboot and doesn't store it anywhere. It's pity they decided not to rely on U-Boot environment for storing configuration.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+/*
+ * Device Tree file for Marvell Armada 385 evaluation board
+ * (DB-88F6820)
```
I see You based this dts on upstream, it should mention this or remove this note.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+/ {
+	model = "WD My Cloud EX2 Ultra";
+	compatible = "marvell,a385-db", "marvell,armada388",
+		 "marvell,armada385", "marvell,armada38x", "marvell,armada380";
```
Add proper compatible string (example: wd,my-cloud-ex2-ultra) and remove "marvell,a385-db", "marvell,armada388".

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+		bootargs = "console=ttyS0,115200";
+		/*
+		append-rootblock = "root=/dev/mtdblock";
+		*/
```
Remove commented node properties.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+			  MBUS_ID(0x0c, 0x04) 0 0xf1200000 0x100000>;
+
+		internal-regs {
+			timer@c200 {
```
All nodes with memory addresses should be sorted from lowest to highest.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+				wp-inverted;
+				bus-width = <8>;
+				status = "disabled";
+				no-1-8-v;
```
If this is disabled don't add properties to this node.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+					reg = <0>;
+					target-supply = <&reg_5v_sata1>;
+				};
+				*/
```
Remove the commented out parts in sata node.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+			};
+
+			/*
+			 * 1GB Flash via NFC NAND controller
```
This seems to not be in line with the specification, does it have 1GB NAND?

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+		regulator-max-microvolt = <12000000>;
+		vin-supply = <&reg_sata1>;
+	};
+*/
```
Is there a need for so much specified regulators which are commented out anyway?

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+*/
+};
+
+&spi0 {
```
I see in boot log:
`m25p80 spi0.0: unrecognized JEDEC id bytes: ff, ff, ff`
seems like it doesn't have SPI memory attached? Does it need to be enabled to properly boot, given the comment at nfc node?

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,419 @@
+	spi-flash@0 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		compatible = "w25q32";
```
If above is true change the compatible string to "jedec,spi-nor".

tmn505 reviewed (commented):
> As far as I understand this device could not be added to the OpenWrt until I send (and kernel upstream accept) patches with the DTS of the device and (at least) wt-poweroff-reset driver.
> 
> And it is not acceptable to have MCU control in a form of a userspace daemon and all its functions should be implemented as drivers (or dropped if it is not feasible) and integrated in kernel as well, right?
> 
> And if kernel won't accept it for some reason — there is no way for the device to get into the OpenWrt distribution, right?
> 
> 
This is not strict requirement and probably will never be, so after reworking the PR to reasonable state it should be accepted. But keep in mind that submitting things to upstream kernel means less work to maintain (if it's accepted), and all devs which have commit rights look favourable on such submissions.

ynezz commented:
> As far as I understand this device could not be added to the OpenWrt until I send (and kernel upstream accept) patches with the DTS of the device and (at least) wt-poweroff-reset driver.

First of all nothing prevents anybody from merging this PR even in it's current state. Review comments are just comments, unless you get explicit NACK, it's always just a view of that particular person and if your view of the things are different, you should simply try to convince the opposite side, that your changes makes sense and should be accepted as they're.

There is no strict rule, which would force you to submit the DTS and that wt-poweroff driver upstream, but I suggest you to do that, as I simply assume, that if you do that, then you've more chances to get this merged.

It's really simple, once this get merged, OpenWrt should support this, so if the DTS/driver stays in the OpenWrt tree and something changes upstream, OpenWrt has to fix this. It's simply taking resources (usually with almost every minor kernel version bump, which happens very frequently, almost bi-weekly) which could be used effectively in other parts of the project. Upstream first has a lot of more advantages as well, code quality is top on this imaginary list.

> And it is not acceptable to have MCU control in a form of a userspace daemon and all its functions should be implemented as drivers (or dropped if it is not feasible) and integrated in kernel as well, right?

In ideal world, that wtmcu package should be optional, not essential in order to get the device into the usable state, so I would try to simply implement the bare minimum needed to boot the kernel and leave the rest in the userspace as optional package (thus avoiding the mandatory usage of the wtmcu stuff in the init scripts), as who needs WOL or some fancy LED patterns in order to boot the device properly, right? Things should be KISS as possible in order to work as expected in most corner cases. The more layers you add, the more could break, rendering the device unusable.

> And if kernel won't accept it for some reason

Well, it depends on the reason, but usually if your initial idea get rejected, you'll get a hint (sometimes several) how it should be implemented properly.

> there is no way for the device to get into the OpenWrt distribution, right?

"No way" are strong words, I would say, that your chances are reduced as merging of this PR (with DTS/wt-poweroff-reset driver in the OpenWrt tree) would lead to more work for OpenWrt folks in the future.

You simply get a lot of bonus points if you're willing to help with the maintenance burden and thus playing nice with the upstream when possible. If the stuff isn't good enough to be accepted upstream, why should it be good enough to be included in the OpenWrt, right?

evgenyz commented:
U-Boot environment:
[uboot-env.txt](https://github.com/openwrt/openwrt/files/3167271/uboot-env.txt)

evgenyz commented:
I cleaned up leftovers, overhauled DTS, joined the device into `cortexa9` subtarget and added `wtmcu` package.

@ynezz 
Thank you for a great explanation. I guess it pretty tiresome to explain all this for every new wanna-be-contributor, I appreciate it.

Regarding the `wtmcu` package. The thing is that LEDs, fan and WOL controls are very meaningful in terms of user experience (or else I won't be bothering with creating it). Here is why: after power up or reset MCU sets status LED to blinking mode (irritating) and fan to maximum speed (10000 rpm, *very* irritating); and there is no power button on the device, so WOL is very convenient to power the device back on after shutdown.

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,297 @@
+
+/ {
+	model = "WD My Cloud EX2 Ultra (BVBZ/Ranger Peak)";
+	compatible = "wd,mcex2u", "wd,armada385", "marvell,armada385", "marvell,armada38x", "marvell,armada380";
```
Remove "wd,armada385" that's not the manufacturer of this processor.
Add new line at end of file.

tmn505 commented on target/linux/mvebu/config-4.14:
```diff
@@ -407,6 +409,9 @@ CONFIG_PM_OPP=y
 CONFIG_PWM_SYSFS=y
+CONFIG_POWER_RESET=y
+# CONFIG_POWER_RESET_QNAP is not set
+CONFIG_POWER_RESET_WTMCU=y
```
I would really like this module to be added as kmod package, but if it's not possible then it's ok.

evgenyz replied:
I'll look into

evgenyz replied:
What if poweroff or reboot occurs before the module is loaded? Board will just be stuck, with the only way out via unplugging the power supply.

evgenyz replied:
You know what, there is very similar driver in kernel `/drivers/power/reset/qnap-poweroff.c`. I'll try to add mine as well (or add a patch to qnap) into the upstream alongside with the DTS. Guess it'd be the best option.

tmn505 replied:
That would be nice.

tmn505 commented on target/linux/mvebu/patches-4.14/531-mvebu-wdmc-libata-leds.patch:
```diff
@@ -0,0 +1,11 @@
++	select ARCH_WANT_LIBATA_LEDS
+ 	select CLKSRC_MMIO
+ 	select PINCTRL
+ 	select PLAT_ORION
```
This patch can be removed if it's already enabled in kernel config, the less patches the better.

evgenyz replied:
Nope. I've tried it. Unless it is present in `menuconfig ARCH_MVEBU` it's got ignored. `target/linux/oxnas` does the same.

tmn505 commented on target/linux/mvebu/image/cortex-a9.mk:
```diff
@@ -183,4 +183,21 @@ define Device/cznic_turris-omnia
+  IMAGE/image.cfs-factory.img := append-ubi
+  IMAGES += uImage-factory.img
+  IMAGE/uImage-factory.img := append-kernel
+  SUPPORTED_DEVICES += armada-385-wd-mcex2u wd,mcex2u
```
You can remove this line. `wd,mcex2u` is set in Default recipe and previously we used dts name as supported devices string, but since Your board was not supported before, there is no need to add it.

evgenyz commented:
I've changed a bit the drivers patch and DTS for them to reflect the version I'm going to try to push upstream. I guess that everything else (except wtmcu package not being a driver) was fixed or proved to be unfixable (like MAC set up).

tmn505 commented:
From my POV this PR looks OK. Thanks for bearing with my comments.

evgenyz commented:
Sure, thank you for your input.

tmn505 commented:
@evgenyz hi again. Could You also add modifications for kernel 4.19 in this PR? Nothing changed drastically, except new NAND driver which uses slightly different dts syntax (You can compare `armada-385-linksys-venom.dts`).

evgenyz commented:
@tmn505 
Done.
Here is also a new bootlog in case I missed something.
[boot-4.19.log](https://github.com/openwrt/openwrt/files/3317298/boot-4.19.log)

adrianschmutzler commented:
Hmm, almost a year since the last comment.

@evgenyz Have you tried to submit upstream as requested (haven't read the whole set of comments)

@tmn505 Can you comment on whether you see a chance for this if it's properly polished (e.g. would you personally invest time to be able to provide a Reviewed-by) or whether we should decline it (i.e. "Won't implement") after almost one year where nobody wanted to review it.

I will go through it and post some comments, whether it's worth addressing them will depend on the answers you give here.

adrianschmutzler commented on package/utils/wtmcu/Makefile:
```diff
@@ -0,0 +1,46 @@
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-v$(PKG_VERSION).tar.bz2
+PKG_SOURCE_URL:=https://gitlab.com/evgenyz/$(PKG_NAME)/-/archive/v$(PKG_VERSION)
```
This doesn't seem to be much code. I'd prefer to have this code in OpenWrt repo instead of having to rely on a private external source with only 2 commits and < 10 files.

evgenyz replied:
Come on, it is GitLab, the largest public GIT repository for open-source projects not owned by Mirosoft. It is hardly private. 90% of OpenWrt packages are like that. I can move it somewhere else (even into the OpenWrt repo) if you like.

It is inactive because it is not used by anything (since this PR is not merged). I expect a bit of action there once it would be used and tested wider. Also, it is not feature-complete in regard of MCU capabilities. Not sure if it would be convenient to work on the daemon if it would be embedded in OpenWrt, but I would rely on your opinion here.

adrianschmutzler replied:
Yes, and if you delete it, it's gone and we have a "broken" device in our tree.
"private" was mostly referring to that it's under your personal control only.

Anyway, again this indicates the disadvantages of relying onto this. Obviously, if this package was moved into OpenWrt, we would have the maintenance burden, and most probably nobody would maintain it if you were occupied otherwise or just wouldn't care anymore in two years (without any offense).

So, as discussed in this PR initially, including this packages really is a major drawback for this proposal, no matter where we put it.

evgenyz replied:
Well, this things could happen, we all mortal in any case.

So I tried my best to make it as harmless as possible. It is BSD-licensed and public. It has almost no dependencies. Chances that it will become broken because of FW update for MCU chip are nearly zero. I also tried to make it as simple as possible while having all the knobs, including ubus support; it'd be easy to take over for the next person interested in this device.

I can transfer it anywhere and make necessary legal modifications for it to be as much independent from me as possible, I have no intention to have a dictatorship over it.

And, of course, right now this device is as broken as it could possibly be because is is not supported at all :)

adrianschmutzler replied:
> And, of course, right now this device is as broken as it could possibly be because is is not supported at all :)

Yes, and as long as we do not support it, we don't have to care :-)

evgenyz replied:
> > And, of course, right now this device is as broken as it could possibly be because is is not supported at all :)
> 
> Yes, and as long as we do not support it, we don't have to care :-)

This is also a legitimate way of thinking. And I fully understand it, despite the fact that it just so happened I have a different one in this particular case :)

adrianschmutzler commented on package/utils/wtmcu/Makefile:
```diff
@@ -0,0 +1,46 @@
+PKG_MAINTAINER:=Evgeny Kolesnikov <evgenyz@gmail.com>
+
+PKG_BUILD_PARALLEL:=1
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-v$(PKG_VERSION)
```
If this becomes a local package, this line can be dropped, so we use the default.

adrianschmutzler commented on package/utils/wtmcu/Makefile:
```diff
@@ -0,0 +1,46 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=wtmcu
```
Creating that package should be a separate commit before adding the actual device.

adrianschmutzler commented on target/linux/mvebu/base-files/lib/preinit/05_wdmc_config:
```diff
@@ -0,0 +1,21 @@
+preinit_mount_config() {
```
base-files should now be created per subtarget, so this should be target/linux/mvebu/cortexaXX/base-files/lib/preinit/05_wdmc_config

adrianschmutzler commented on target/linux/mvebu/base-files/lib/preinit/06_set_iface_mac:
```diff
@@ -40,6 +40,14 @@ preinit_set_mac_address() {
 		;;
+	wd,mcex2u)
+		if [ -f /tmp/wd-config/mac_addr ]; then
+			mac=$(cat /tmp/wd-config/mac_addr)
```
So, what's the condition of this file to be available?

evgenyz replied:
It is a configuration file from the vendor config partition being mounted in `preinit_mount_config`. It might be corrupted or removed, so the script checks if it is present before trying to fetch the MAC.

adrianschmutzler replied:
okay, so there is a legitimate reason why the file wouldn't be there. I just want to make sure there is a reason behind these kind of checks...

However, don't think checking with -f won't help much if the file is corrupted :-)

evgenyz replied:
It is for the case when the config filesystem is corrupted and could not be mounted.

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+
+/ {
+	model = "WD My Cloud EX2 Ultra (BVBZ/Ranger Peak)";
+	compatible = "wd,mcex2u", "marvell,armada385", "marvell,armada38x",
```
I'm not a fan of these abbreviations. Recently, we frequently just took the device name instead, creating additional consistency of device names:
wd,my-cloud-ex2-ultra

Please update compatible, board-files, DTS file name and Makefile accordingly.

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+
+	memory {
+		device_type = "memory";
+		reg = <0x00000000 0x40000000>; /* 1024 MB */
```
do we really need this comment

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+	soc {
+		ranges = <MBUS_ID(0xf0, 0x01) 0 0xf1000000 0x100000
+			  MBUS_ID(0x01, 0x1d) 0 0xfff00000 0x100000
+			  MBUS_ID(0x09, 0x19) 0 0xf1100000 0x10000
```
It would be nicer to add a leading zero here, i.e. 0x010000

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+					label = "U-Boot";
+					reg = <0x00000000 0x00500000>;
+					read-only;
+				};
```
Add empty lines between partition nodes.

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+					reg = <0x00000000 0x00500000>;
+					read-only;
+				};
+				partition@0500000 {
```
I'd remove the leading zeros in the partition@ names, i.e. partition@0, partition@500000, etc.

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+		pinctrl-0 = <&sata_leds>;
+
+		sata1-red-led {
+			label = "wdmcex2u:red:hdd1";
```
The first part should match the model part of the compatible, i.e. my-cloud-ex2-ultra

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+		pinctrl-names = "default";
+		pinctrl-0 = <&sata_leds>;
+
+		sata1-red-led {
```
led suffix is unnecessary, despite I'd be consistent with the label, thus I'd choose "hdd1_red"

adrianschmutzler commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+		sata1-red-led {
+			label = "wdmcex2u:red:hdd1";
+			gpios = <&gpio1 11 GPIO_ACTIVE_HIGH>;
+		};
```
Empty lines between nodes.

adrianschmutzler commented on target/linux/mvebu/files-4.19/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Device Tree file for Western Digital My Cloud EX2 Ultra
+ * (BVBZ/Ranger Peak)
```
Same comments as for version above apply.

adrianschmutzler commented on target/linux/mvebu/image/cortex-a9.mk:
```diff
@@ -183,4 +183,20 @@ define Device/cznic_turris-omnia
 
+define Device/wd_mcex2u
+  $(Device/NAND-128K)
+  DEVICE_TITLE := WD My Cloud EX2 Ultra (BVBZ/Ranger Peak)
```
Replace by recent DEVICE_VENDOR/DEVICE_MODEL syntax.

adrianschmutzler commented on target/linux/mvebu/image/cortex-a9.mk:
```diff
@@ -183,4 +183,20 @@ define Device/cznic_turris-omnia
+  KERNEL_SIZE := 5120k
+  IMAGES += image.cfs-factory.img
+  IMAGE/image.cfs-factory.img := append-ubi
+  IMAGES += uImage-factory.img
```
I'd prefer a more consistent scheme for naming these images.

evgenyz replied:
They are named after the partitions on the device (original DTS) in order to help with installation. What can you propose for better names?

evgenyz replied:
```
0x000000000000-0x000000500000 : "U-Boot"
0x000000500000-0x000000a00000 : "uImage"
0x000000a00000-0x000000f00000 : "uRamdisk"
0x000000f00000-0x00000c800000 : "image.cfs"
0x00000c800000-0x00000d700000 : "rescue fw"
0x00000d700000-0x00000eb00000 : "config"
0x00000eb00000-0x00000f500000 : "reserve1"
0x00000f500000-0x00000ff00000 : "reserve2"
```

adrianschmutzler replied:
Ah, now I see. I really don't like the dot for the first one, as it looks like you apply two different styles here.

I'd consider just replacing the dot by a hyphen, which will preserve your initial idea but won't look so strange IMO:
`IMAGE/image-cfs-factory.img`

Despite, I'd add the IMAGES in one line, followed by the definitions:
```
IMAGES += image-cfs-factory.img uImage-factory.img
IMAGE/image-cfs-factory.img := ...
IMAGE/uImage-factory.img := ...
```

I'd personally prefer .bin to .img, but I'm not sure what's common here on mvebu.

adrianschmutzler commented:
Needs to be updated for kernel 5.4, 4.14 can be dropped.

After quick review, this looks like it still needs quite some work from my perspective, where the chances for being merged are still low/unclear.

evgenyz commented on target/linux/mvebu/base-files/etc/diag.sh:
```diff
@@ -28,7 +28,19 @@ get_status_led() {
 }
 
-set_state() {
+set_state_wtmcu() {
```
Huh, this is a nice throw of a wrench :(

evgenyz replied:
Okay, I'm kinda lost here. Is there a legitimate way to override `base-files` on per-device level (build-time or run-time), I can't find anything useful :( @tmn505, maybe you could give some advice?

tmn505 replied:
There's no other option as to extend `diag.sh` to load/execute additional functions/cases from (example) `/lib/functions/platform.sh` which could be overridden per target.

evgenyz commented:
Yeah, time flies. I tried to submit the driver patch and DTS [to the upstream](https://patchwork.kernel.org/cover/11053373/) and after some discussion we had sort of an agreement on how the change should look like. 

Then some guy contacted me about the implementation of a full-wrap driver set for this MCU controller (which could have opened the possibility of getting rid of the wtmcud daemon). Never heard from him since. And then I myself got carried away.

So, most of the comment you made are pretty trivial and I can fix it easily. I can also again push for the driver and DTS to be included in upstream with all the required changes, that is most likely doable as well.

There is only one major roadblock here — the wtmcu package/daemon. I can't promise that I could be able to invest adequate amount of effort to move it into drivers space. If this is acceptable, I can fix everything else.

tmn505 commented:
> @tmn505 Can you comment on whether you see a chance for this if it's properly polished (e.g. would you personally invest time to be able to provide a Reviewed-by) or whether we should decline it (i.e. "Won't implement") after almost one year where nobody wanted to review it.

Given my knowledge at that time, it looked fine after my initial review https://github.com/openwrt/openwrt/pull/2040#issuecomment-493975475 (should have added "Approved" in github interface), but also @ynezz added "needs changes" label and never removed it, so I assume(d) there's something he thinks was not addressed after his comments.


> Yeah, time flies.

Indeed.

> I tried to submit the driver patch and DTS [to the upstream](https://patchwork.kernel.org/cover/11053373/) and after some discussion we had sort of an agreement on how the change should look like.
> 
> Then some guy contacted me about the implementation of a full-wrap driver set for this MCU controller (which could have opened the possibility of getting rid of the wtmcud daemon). Never heard from him since. And then I myself got carried away.
> 
> So, most of the comments you made are pretty trivial and I can fix these problems easily. I can also again push for the driver and DTS to be included in upstream with all the required changes, that is most likely doable as well.

Yes, please do, that always lifts load from devs backs. I'll also post few comments here, mostly if not only, stylistic ones.

> 
> There is only one major roadblock here — the wtmcu package/daemon. I can't promise that I could be able to invest adequate amount of effort to move it into drivers space. If this is acceptable, I can fix everything else.

Until it works with driver implementation accepted upstream, I think it's acceptable in whatever form You see fit (personal opinion).

tmn505 commented on target/linux/mvebu/files-4.14/arch/arm/boot/dts/armada-385-wd-mcex2u.dts:
```diff
@@ -0,0 +1,306 @@
+				status = "disabled";
+			};
+
+			i2c0: i2c@11000 {
```
If the node from "internal-regs" has a label in "armada-38x.dtsi" please refer the node by its label outside of root node.

tmn505 reviewed (commented):
Please also address the upstream comments.

adrianschmutzler commented:
I just read through this thing again, and my conclusion is effectively the consequence of what ynezz predicted relatively early: the major problems seem to arise from the wtmcu package being included here.

If it is removed from this PR, most of our discussion points will vanish. The device will still boot (of course in a more inferior state). I'd go that way, and there will still be enough work left ...

evgenyz commented:
> I just read through this thing again, and my conclusion is effectively the consequence of what ynezz predicted relatively early: the major problems seem to arise from the wtmcu package being included here.
> If it is removed from this PR, most of our discussion points will vanish. The device will still boot (of course in a more inferior state). I'd go that way, and there will still be enough work left ...

I already fixed most of them, working on drivers now and trying to do something with `diags.sh`, without re-introducing it for the whole `a9`. I guess that if we will make `wtmcu` package optional and move it to `packages`, there would be even less opportunities to influence `base-files`.

evgenyz commented:
My current ideas: 
- add more flexibility to `base-files` Makefile in order to match granularity level with the whole build system (all way down to TARGET_DEVICE);
- add more flexibility to `diag.sh` (not sure how yet, but I guess this most likely will be rejected).

adrianschmutzler commented:
Just be aware that my comment after almost one year of inactivity was: Does anybody see a chance that this will be reviewed or can we close it right away?

So, the most probable outcome will be that you invest a lot of time to polish this and then nobody will pick it up, like before. Just add that to your considerations.

Don't feel repelled, it's just the limited manpower of a FOSS project ...

tmn505 commented:
@adrianschmutzler I added my comments and will review after new changes will be pushed.

@evgenyz go for adding more flexibility to `diag.sh`, I don't think it'll be rejected.

evgenyz commented:
Fresh [boot-5.4.log](https://github.com/openwrt/openwrt/files/4706149/boot-5.4.log).

The `diag.sh` script could be simplified when (if) #3046 would be merged.

@tmn505, @adrianschmutzler I guess I addressed all your comments as well as comments from upstream. The reboot and power-off drivers are now a patch against QNAP Power off driver, this is how upstream prefer them to be (not yet accepted).

tmn505 reviewed (approved):
Aside from `diag.sh` changes, which are still discussed, everything looks good.

evgenyz commented:
Rebased.

adschm commented:
After almost another year, I think it's finally time to put this to rest. Sorry.
